content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,384 @@
+      updatedState.derivedBar = deriveBar(props.bar);
+    }
+    if (props.foo !== state.prevFooFromProps) {
+      updatedState.prevFooFromProps = props.foo;
+      updatedState.derivedFoo = deriveFoo(props.foo);
+    }
+    // Other derived fields would go here...
+    return updatedState;
+  }
+
+  // ...
+}
+```
+
+Although the above pattern _works_, it isn't what `getDerivedStateFromProps` is meant to be used for. This component isn't reacting to **changes in props**. It's just computing values based on the **current props**, and there is a simpler way to do this that doesn't even require state.
+
+Let's look at another example to illustrate this. Here is a component that accepts an array of items as `props.list` and allows a user to filter the array by entering text. We could use `getDerivedStateFromProps` for this purpose:
+
+```js
+class Example extends Component {
+  state = {
+    filterText: "",
+    prevFilterText: ""
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Re-run the filter whenever the list array or filter text change:
+    if (
+      props.list !== state.prevPropsList ||
+      state.prevFilterText !== state.filterText
+    ) {
+      // Note that to do this, state needs to explicitly track both:
+      // prevPropsList - so we can detect a change in props.list
+      // prevFilterText - so we can detect a change in state.filterText
+      return {
+        prevPropsList: props.list,
+        prevFilterText: state.filterText,
+        filteredList: props.list.filter(item => item.text.includes(state.filterText))
+      };
+    }
+    return null;
+  }
+
+  render() {
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{this.state.filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => {
+    this.setState({ filterText: event.target.value });
+  };
+}
+```
+
+This implementation is more complicated than it needs to be, because it has to separately track and detect changes in both props and state in order to properly update the filtered list. In this example, we could simplify things by just using `PureComponent` and moving the filter operation into the render method: 
+
+```js
+// PureComponents only re-render if at least one state or prop value changes.
+// Change is determined by doing a shallow comparison of state and prop keys.
+class Example extends PureComponent {
+  // State only needs to hold the current filter text value:
+  state = {
+    filterText: ""
+  };
+
+  render() {
+    const filteredList = this.props.list.filter(
+      item => item.text.includes(this.state.filterText)
+    )
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => {
```

Comment:
Nit (in this and other examples): I think most or maybe all doc examples put event handlers before render so it jumps out a little bit.